From 0d81710b6ffe5bf9c65ea10a959c16a8f51ca5a0 Mon Sep 17 00:00:00 2001
From: Jack Tan <jack.tan@windriver.com>
Date: Wed, 25 May 2011 15:52:33 +0800
Subject: [PATCH] nlm_xls_atx_64_be: fixed the definition of cache coherency attribute bit

SDK 1.7 use the fixed offset value in the three cache attributs macro
It's not correct when we disable the HUGETLB_PAGE.

The _CACHE_CACHABLE_NONCOHERENT affect the PAGE_NONE macro in protection_map[]
which is the mapping table of vm protection attributes and hardware protection
attributes. Kernel would get the incorrect attribute bits when using the table.

It can make the cache attributs to 0x7 when kernel modify the protection bits
by mprotect() system call. But 0x7 is illegal in XLS/XLR platform and it will
cause Address Error Exception:

Unhandled kernel unaligned access[#1]:
Cpu 0
$ 0   : 0000000000000000 000000001000dce0 0000000000000000 ffffff0000000000
......
......
$16   : a80000008d617eb0 ffffff0000000000 ffffffff8e050000 000000002aace000
$20   : 000000000000118e 0000000000000000 00000000004d2ab8 00000000004a0000
$24   : ffffffffac000000 000000002ae2b84c
$28   : a80000008d614000 a80000008d617e80 00000000004c0000 ffffffff83401dc4
Hi    : 0000000000000000
Lo    : 00000082d126a800
epc   : ffffffff83425c28 do_ade+0x3a8/0x5d0
    Tainted: G        W
ra    : ffffffff83401dc4 ret_from_exception+0x0/0x10
Status: 1000dce3    KX SX UX KERNEL EXL IE
Cause : 00800010
BadVA : 000000002aace000
PrId  : 000c4402 (XLS416 Rev B0)
Modules linked in:
Process vsftpd (pid: 298, threadinfo=a80000008d614000, task=a80000008df86890...)
Stack : 000000007fdc8498 000000002aace000 000000007fdc84e8 000000000000118e
        000000007fdc8498 ffffffff83401dc4 0000000000000000 000000001000dce0
        ...
        ...
Call Trace:
[<ffffffff83425c28>] do_ade+0x3a8/0x5d0
[<ffffffff83401dc4>] ret_from_exception+0x0/0x10

Signed-off-by: Jack Tan <jack.tan@windriver.com>
---
 arch/mips/include/asm/pgtable-bits.h |    6 +++---
 1 files changed, 3 insertions(+), 3 deletions(-)

diff --git a/arch/mips/include/asm/pgtable-bits.h b/arch/mips/include/asm/pgtable-bits.h
index cccecd0..fdd3784 100644
--- a/arch/mips/include/asm/pgtable-bits.h
+++ b/arch/mips/include/asm/pgtable-bits.h
@@ -194,9 +194,9 @@ static inline uint64_t pte_to_entrylo(unsigned long pte_val)
 
 #elif defined(CONFIG_CPU_PHOENIX)
 
-#define _CACHE_UNCACHED             (2<<9)
-#define _CACHE_CACHABLE_COW         (3<<9)
-#define _CACHE_CACHABLE_NONCOHERENT (3<<9)
+#define _CACHE_UNCACHED             (2<<_CACHE_SHIFT)
+#define _CACHE_CACHABLE_COW         (3<<_CACHE_SHIFT)
+#define _CACHE_CACHABLE_NONCOHERENT (3<<_CACHE_SHIFT)
 
 #elif defined(CONFIG_CPU_RM9000)
 
-- 
1.6.5.2

